File: backend/src/routes/locations.ts
import { Hono } from 'hono';
import { z } from 'zod';
import { zValidator } from '@hono/zod-validator';
import type { JSONValue } from '../lib/db.ts';
import { requireAuth, getUser, type AppEnv } from '../middleware/auth.ts';
import { withUserDb, withAnonDb } from '../middleware/rls.ts';
import { Forbidden, NotFound } from '../lib/errors.ts';
import { bootstrapPersonalAccount, makeLocationSlug } from './auth.ts';
import { DAY_S, fixedWindowStart } from '../lib/rate-limit.ts';

const createLocationSchema = z
  .object({
    parent_location_id: z.string().uuid().nullable().optional(),
    type: z.enum(['house', 'complex', 'building', 'other']),
    name: z.string().min(1).max(120),
    slug: z.string().min(1).max(120).optional(),
    address: z.record(z.unknown()).optional(),
    lat: z.number().optional(),
    long: z.number().optional(),
  })
  .strict();

// New top-level create: locations are first-class. Each one gets a fresh
// account (1:1) so its members are isolated from any other location the
// same user owns.
const createTopLevelLocationSchema = z
  .object({
    name: z.string().min(1).max(120),
    type: z.enum(['house', 'complex', 'building', 'other']).default('house'),
    country_code: z
      .string()
      .length(2)
      .transform((v) => v.toUpperCase())
      .default('ZA'),
    address: z.record(z.unknown()).optional(),
    lat: z.number().optional(),
    long: z.number().optional(),
  })
  .strict();

// Abuse-protection quotas (NOT billing — whatsacc has none). NULL clears a
// cap (= unlimited); omitted fields are left unchanged.
const patchLimitsSchema = z
  .object({
    max_opens_per_member_per_day: z.number().int().min(1).max(100_000).nullable().optional(),
    max_opens_per_location_per_day: z.number().int().min(1).max(1_000_000).nullable().optional(),
  })
  .strict();

const patchLocationSchema = z
  .object({
    name: z.string().min(1).max(120).optional(),
    address: z.record(z.unknown()).optional(),
    lat: z.number().optional(),
    long: z.number().optional(),
    status: z.string().optional(),
  })
  .strict();

function locationsRouter() {
  const app = new Hono<AppEnv>();
  app.use('*', requireAuth());

  app.get('/accounts/:accountId/locations', async (c) => {
    const accountId = c.req.param('accountId');
    const rows = await withUserDb(c, async (tx) => {
      return await tx<{
        id: string;
        parent_location_id: string | null;
        type: string;
        name: string;
        slug: string | null;
        status: string;
        address: unknown;
        access_point_count: string;
        member_count: string;
        last_opened_at: Date | null;
      }[]>`
        select
          l.id, l.parent_location_id, l.type, l.name, l.slug, l.status, l.address,
          (select count(*)::text from access_points ap where ap.location_id = l.id) as access_point_count,
          (select count(*)::text from location_members lm where lm.location_id = l.id) as member_count,
          (select max(al.ts) from access_logs al
             where al.location_id = l.id and al.command = 'open' and al.success = true) as last_opened_at
        from locations l
        where l.account_id = ${accountId}
        order by l.created_at asc
      `;
    });
    return c.json({
      locations: rows.map((r) => ({
        id: r.id,
        parent_location_id: r.parent_location_id,
        type: r.type,
        name: r.name,
        slug: r.slug,
        status: r.status,
        address: r.address,
        access_point_count: Number(r.access_point_count),
        member_count: Number(r.member_count),
        last_opened_at: r.last_opened_at,
      })),
    });
  });

  app.post(
    '/accounts/:accountId/locations',
    zValidator('json', createLocationSchema),
    async (c) => {
      const user = getUser(c);
      const accountId = c.req.param('accountId');
      const body = c.req.valid('json');
      const result = await withUserDb(c, async (tx) => {
        const rows = await tx<{ id: string }[]>`
          insert into locations
            (account_id, parent_location_id, type, name, slug, address, lat, long, status)
          values
            (${accountId}, ${body.parent_location_id ?? null}, ${body.type},
             ${body.name}, ${body.slug ?? makeLocationSlug(body.name)}, ${tx.json((body.address ?? {}) as JSONValue)},
             ${body.lat ?? null}, ${body.long ?? null}, 'active')
          returning id
        `;
        await tx`
          insert into location_members (location_id, user_id, role)
          values (${rows[0]!.id}, ${user.sub}, 'owner')
          on conflict (location_id, user_id) do update set role = excluded.role, updated_at = now()
        `;
        return rows[0]!;
      });
      return c.json(result, 201);
    },
  );

  // -------------------------------------------------------------------------
  // Abuse-protection quotas ("limits") + usage. GET is member-visible so the
  // portal can show "3 of 4 opens used"; PATCH is admin-only. Usage is
  // computed from access_logs over the same UTC day window the enforcement
  // counters use, so what the UI shows matches what the limiter counts.
  // -------------------------------------------------------------------------

  app.get('/:id/limits', async (c) => {
    const id = c.req.param('id');
    const user = getUser(c);
    const dayStart = fixedWindowStart(new Date(), DAY_S);
    const data = await withUserDb(c, async (tx) => {
      const loc = await tx<{ id: string; account_id: string }[]>`
        select id, account_id from locations where id = ${id}
      `;
      if (!loc[0]) throw NotFound('location_not_found');

      const settings = await tx<
        { max_opens_per_member_per_day: number | null; max_opens_per_location_per_day: number | null }[]
      >`
        select max_opens_per_member_per_day, max_opens_per_location_per_day
        from location_settings where location_id = ${id}
      `;

      const totals = await tx<{ location_opens_today: string; my_opens_today: string }[]>`
        select
          count(*) filter (where success = true and command = 'open')::text as location_opens_today,
          count(*) filter (
            where success = true and command = 'open' and user_id = ${user.sub}
          )::text as my_opens_today
        from access_logs
        where location_id = ${id} and ts >= ${dayStart}
      `;

      // Per-member breakdown for the admin UI. Email is left-joined under
      // the caller's RLS — other users' rows resolve to null email, which
      // the frontend maps via the members list.
      const members = await tx<{ user_id: string | null; email: string | null; opens_today: string }[]>`
        select al.user_id, u.email::text as email, count(*)::text as opens_today
        from access_logs al
        left join users u on u.id = al.user_id
        where al.location_id = ${id}
          and al.success = true
          and al.command = 'open'
          and al.ts >= ${dayStart}
        group by al.user_id, u.email
        order by count(*) desc
        limit 50
      `;

      return {
        quotas: {
          max_opens_per_member_per_day: settings[0]?.max_opens_per_member_per_day ?? null,
          max_opens_per_location_per_day: settings[0]?.max_opens_per_location_per_day ?? null,
        },
        usage: {
          day_start: dayStart.toISOString(),
          location_opens_today: Number(totals[0]?.location_opens_today ?? 0),
          my_opens_today: Number(totals[0]?.my_opens_today ?? 0),
          members: members.map((m) => ({
            user_id: m.user_id,
            email: m.email,
            opens_today: Number(m.opens_today),
          })),
        },
      };
    });
    return c.json({ location_id: id, ...data });
  });

  app.patch('/:id/limits', zValidator('json', patchLimitsSchema), async (c) => {
    const id = c.req.param('id');
    const body = c.req.valid('json');
    const hasMember = 'max_opens_per_member_per_day' in body;
    const hasLocation = 'max_opens_per_location_per_day' in body;
    const memberVal = body.max_opens_per_member_per_day ?? null;
    const locationVal = body.max_opens_per_location_per_day ?? null;

    const quotas = await withUserDb(c, async (tx) => {
      // Pre-check admin so the caller gets a clean 403/404 instead of an
      // RLS-shaped 500 on the location_settings upsert.
      const loc = await tx<{ id: string; admin: boolean }[]>`
        select l.id, app.is_account_admin(l.account_id) as admin
        from locations l where l.id = ${id}
      `;
      if (!loc[0]) throw NotFound('location_not_found');
      if (!loc[0].admin) throw Forbidden('not_account_admin');

      const rows = await tx<
        { max_opens_per_member_per_day: number | null; max_opens_per_location_per_day: number | null }[]
      >`
        insert into location_settings
          (location_id, max_opens_per_member_per_day, max_opens_per_location_per_day)
        values (${id}, ${memberVal}, ${locationVal})
        on conflict (location_id) do update set
          max_opens_per_member_per_day = case
            when ${hasMember} then ${memberVal}::int
            else location_settings.max_opens_per_member_per_day end,
          max_opens_per_location_per_day = case
            when ${hasLocation} then ${locationVal}::int
            else location_settings.max_opens_per_location_per_day end,
          updated_at = now()
        returning max_opens_per_member_per_day, max_opens_per_location_per_day
      `;
      return rows[0]!;
    });
    return c.json({ location_id: id, quotas });
  });

  app.get('/:id', async (c) => {
    const id = c.req.param('id');
    const data = await withUserDb(c, async (tx) => {
      const rows = await tx<{
        id: string;
        account_id: string;
        parent_location_id: string | null;
        type: string;
        name: string;
        slug: string | null;
        address: unknown;
        lat: number | null;
        long: number | null;
        status: string;
      }[]>`
        select id, account_id, parent_location_id, type, name, slug, address, lat, long, status
        from locations where id = ${id}
      `;
      return rows[0] ?? null;
    });
    if (!data) throw NotFound('location_not_found');
    return c.json(data);
  });

  // Top-level POST /locations — creates a fresh account + location pair
  // owned by the caller. Used when a user adds a NEW location ("invite my
  // cleaner to a different house"); each one gets isolated members. The
  // caller is added as 'owner' of the new account.
  app.post('/', zValidator('json', createTopLevelLocationSchema), async (c) => {
    const user = getUser(c);
    const body = c.req.valid('json');
    const result = await withAnonDb(async (tx) => {
      const accountId = await bootstrapPersonalAccount(tx, {
        userId: user.sub,
        name: body.name,
        countryCode: body.country_code,
      });
      const rows = await tx<{ id: string }[]>`
        insert into locations
          (account_id, type, name, slug, address, lat, long, status)
        values (
          ${accountId},
          ${body.type},
          ${body.name},
          ${makeLocationSlug(body.name)},
          ${tx.json((body.address ?? {}) as JSONValue)},
          ${body.lat ?? null},
          ${body.long ?? null},
          'active'
        )
        returning id
      `;
      await tx`
        insert into location_members (location_id, user_id, role)
        values (${rows[0]!.id}, ${user.sub}, 'owner')
        on conflict (location_id, user_id) do update set role = excluded.role, updated_at = now()
      `;
      return { id: rows[0]!.id, account_id: accountId };
    });
    return c.json(result, 201);
  });

  // DELETE /locations/:id — drops the location and (since 1:1) its parent
  // account if no sibling locations remain. The cascade chain handles
  // members, devices, access_points etc. RLS guarantees only an owner/admin
  // of the parent account can do this.
  app.delete('/:id', async (c) => {
    const result = await withUserDb(c, async (tx) => {
      const found = await tx<{ id: string; account_id: string }[]>`
        select id, account_id from locations where id = ${c.req.param('id')}
      `;
      if (!found[0]) throw NotFound('location_not_found');
      const accountId = found[0].account_id;
      await tx`delete from locations where id = ${found[0].id}`;
      // If this account has no other locations, drop the account too so
      // the user isn't left with an orphaned tenant.
      const remaining = await tx<{ count: string }[]>`
        select count(*)::text from locations where account_id = ${accountId}
      `;
      if (remaining[0] && Number(remaining[0].count) === 0) {
        await tx`delete from accounts where id = ${accountId}`;
      }
      return { deleted: found[0].id, account_dropped: Number(remaining[0]?.count ?? 1) === 0 };
    });
    return c.json(result);
  });

  app.patch('/:id', zValidator('json', patchLocationSchema), async (c) => {
    const id = c.req.param('id');
    const body = c.req.valid('json');
    // TODO: build dynamic update; for now do a single set with COALESCEs.
    // Use RETURNING to detect when RLS filtered the update to zero rows
    // (caller is not a member of the owning account) — without this we'd
    // return 204 even though nothing changed, which leaks "the row exists"
    // and confuses the caller into thinking their PATCH applied.
    await withUserDb(c, async (tx) => {
      const updated = await tx<{ id: string }[]>`
        update locations set
          name = coalesce(${body.name ?? null}, name),
          address = coalesce(${tx.json((body.address ?? null) as JSONValue)}, address),
          lat = coalesce(${body.lat ?? null}, lat),
          long = coalesce(${body.long ?? null}, long),
          status = coalesce(${body.status ?? null}, status),
          updated_at = now()
        where id = ${id}
        returning id
      `;
      if (updated.length === 0) throw NotFound('location_not_found');
    });
    return c.body(null, 204);
  });

  return app;
}

export const locationsRoutes = locationsRouter();

File: gateway/internal/store/locations.go
package store

import (
	"context"
	"database/sql"
)

// LocationDetail is the GET /locations shape: the row plus address json and
// the aggregate counts the portal shows.
type LocationDetail struct {
	Location
	ParentLocationID string
	Address          string // raw json object, "{}" default
	Lat, Long        sql.NullFloat64
	AccessPointCount int
	MemberCount      int
	LastOpenedAt     sql.NullInt64
}

// LocationsByAccountDetailed lists the account's locations with counts +
// last successful open, mirroring GET /accounts/:id/locations.
func (s *Store) LocationsByAccountDetailed(ctx context.Context, accountID string) ([]LocationDetail, error) {
	rows, err := s.db.QueryContext(ctx,
		`SELECT l.id, l.account_id, coalesce(l.parent_location_id, ''), l.type, l.name, l.slug,
		        l.status, l.address, l.lat, l.long,
		        (SELECT count(*) FROM access_points ap WHERE ap.location_id = l.id),
		        (SELECT count(*) FROM location_members lm WHERE lm.location_id = l.id),
		        (SELECT max(al.ts) FROM access_logs al
		          WHERE al.location_id = l.id AND al.command = 'open' AND al.success = 1)
		 FROM locations l
		 WHERE l.account_id = ?
		 ORDER BY l.created_at ASC`, accountID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []LocationDetail
	for rows.Next() {
		var d LocationDetail
		if err := rows.Scan(&d.ID, &d.AccountID, &d.ParentLocationID, &d.Type, &d.Name, &d.Slug,
			&d.Status, &d.Address, &d.Lat, &d.Long,
			&d.AccessPointCount, &d.MemberCount, &d.LastOpenedAt); err != nil {
			return nil, err
		}
		out = append(out, d)
	}
	return out, rows.Err()
}

// LocationDetailByID fetches one location (with address/coords) iff it
// belongs to accountID.
func (s *Store) LocationDetailByID(ctx context.Context, accountID, id string) (*LocationDetail, error) {
	var d LocationDetail
	err := s.db.QueryRowContext(ctx,
		`SELECT id, account_id, coalesce(parent_location_id, ''), type, name, slug, status, address, lat, long
		 FROM locations WHERE id = ? AND account_id = ?`, id, accountID).
		Scan(&d.ID, &d.AccountID, &d.ParentLocationID, &d.Type, &d.Name, &d.Slug, &d.Status,
			&d.Address, &d.Lat, &d.Long)
	if err != nil {
		return nil, err
	}
	return &d, nil
}

// CreateLocationArgs are the nested-create inputs (POST /accounts/:id/locations).
type CreateLocationArgs struct {
	ParentLocationID string
	Type             string
	Name             string
	Slug             string // "" = derive from name
	AddressJSON      string // "" = "{}"
	Lat, Long        *float64
	CreatorUserID    string // added as location 'owner' member
}

// CreateLocationFull inserts a location under the account and makes the
// creator a location owner-member.
func (s *Store) CreateLocationFull(ctx context.Context, accountID string, a CreateLocationArgs) (string, error) {
	slug := a.Slug
	if slug == "" {
		slug = Slugify(a.Name)
		if slug == "" {
			slug = "home"
		}
	}
	addr := a.AddressJSON
	if addr == "" {
		addr = "{}"
	}
	t := now()
	id := NewID()
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return "", err
	}
	defer tx.Rollback()
	if _, err := tx.ExecContext(ctx,
		`INSERT INTO locations (id, account_id, parent_location_id, type, name, slug, address, lat, long, status, created_at, updated_at)
		 VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'active', ?, ?)`,
		id, accountID, nullable(a.ParentLocationID), a.Type, a.Name, slug, addr,
		nullFloat(a.Lat), nullFloat(a.Long), t, t); err != nil {
		return "", err
	}
	if a.CreatorUserID != "" {
		if _, err := tx.ExecContext(ctx,
			`INSERT INTO location_members (location_id, user_id, role, created_at, updated_at)
			 VALUES (?, ?, 'owner', ?, ?)
			 ON CONFLICT (location_id, user_id) DO UPDATE SET role = 'owner', updated_at = excluded.updated_at`,
			id, a.CreatorUserID, t, t); err != nil {
			return "", err
		}
	}
	return id, tx.Commit()
}

// LocationAccountID resolves a location to its owning account — the gating
// lookup handlers use before a MemberRole check (never exposed raw).
func (s *Store) LocationAccountID(ctx context.Context, id string) (string, error) {
	var accountID string
	err := s.db.QueryRowContext(ctx, `SELECT account_id FROM locations WHERE id = ?`, id).Scan(&accountID)
	return accountID, err
}

// UpdateLocationType changes the location type (top-level create adjusts the
// anchor location after CreateAccountWithOwner defaults it to 'house').
func (s *Store) UpdateLocationType(ctx context.Context, accountID, id, typ string) error {
	res, err := s.db.ExecContext(ctx,
		`UPDATE locations SET type = ?, updated_at = ? WHERE id = ? AND account_id = ?`,
		typ, now(), id, accountID)
	if err != nil {
		return err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	return nil
}

// LocationPatch is the PATCH /locations/:id partial update. Nil = unchanged.
type LocationPatch struct {
	Name        *string
	AddressJSON *string
	Lat, Long   *float64
	Status      *string
}

// UpdateLocation applies the patch iff the location belongs to accountID
// (ErrNotFound otherwise — cross-tenant PATCH must not leak existence).
func (s *Store) UpdateLocation(ctx context.Context, accountID, id string, p LocationPatch) error {
	res, err := s.db.ExecContext(ctx,
		`UPDATE locations SET
		   name    = coalesce(?, name),
		   address = coalesce(?, address),
		   lat     = coalesce(?, lat),
		   long    = coalesce(?, long),
		   status  = coalesce(?, status),
		   updated_at = ?
		 WHERE id = ? AND account_id = ?`,
		nullStr(p.Name), nullStr(p.AddressJSON), nullFloat(p.Lat), nullFloat(p.Long), nullStr(p.Status),
		now(), id, accountID)
	if err != nil {
		return err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return ErrNotFound
	}
	return nil
}

// DeleteLocation drops the location and — locations being 1:1 with accounts
// in the product model — the parent account when no sibling remains.
// Returns whether the account was dropped too.
func (s *Store) DeleteLocation(ctx context.Context, accountID, id string) (accountDropped bool, err error) {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return false, err
	}
	defer tx.Rollback()
	res, err := tx.ExecContext(ctx,
		`DELETE FROM locations WHERE id = ? AND account_id = ?`, id, accountID)
	if err != nil {
		return false, err
	}
	if n, _ := res.RowsAffected(); n == 0 {
		return false, ErrNotFound
	}
	var remaining int
	if err := tx.QueryRowContext(ctx,
		`SELECT count(*) FROM locations WHERE account_id = ?`, accountID).Scan(&remaining); err != nil {
		return false, err
	}
	if remaining == 0 {
		if _, err := tx.ExecContext(ctx, `DELETE FROM accounts WHERE id = ?`, accountID); err != nil {
			return false, err
		}
	}
	return remaining == 0, tx.Commit()
}

// ---------------------------------------------------------------------------
// Quotas ("limits") + usage
// ---------------------------------------------------------------------------

// Quotas are the two optional per-location daily caps. Nil = unlimited.
type Quotas struct {
	MaxOpensPerMemberPerDay   *int64
	MaxOpensPerLocationPerDay *int64
}

// LocationQuotas reads the location's quota settings (zero-value Quotas when
// no settings row exists).
func (s *Store) LocationQuotas(ctx context.Context, locationID string) (Quotas, error) {
	var q Quotas
	var m, l sql.NullInt64
	err := s.db.QueryRowContext(ctx,
		`SELECT max_opens_per_member_per_day, max_opens_per_location_per_day
		 FROM location_settings WHERE location_id = ?`, locationID).Scan(&m, &l)
	if err == sql.ErrNoRows {
		return q, nil
	}
	if err != nil {
		return q, err
	}
	if m.Valid {
		q.MaxOpensPerMemberPerDay = &m.Int64
	}
	if l.Valid {
		q.MaxOpensPerLocationPerDay = &l.Int64
	}
	return q, nil
}

// PatchLocationQuotas upserts location_settings with omitted-vs-null
// semantics: hasX=false leaves the column unchanged, hasX=true sets it to
// val (nil = clear = unlimited). Returns the resulting quotas.
func (s *Store) PatchLocationQuotas(ctx context.Context, locationID string, hasMember bool, memberVal *int64, hasLocation bool, locationVal *int64) (Quotas, error) {
	t := now()
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return Quotas{}, err
	}
	defer tx.Rollback()
	if _, err := tx.ExecContext(ctx,
		`INSERT INTO location_settings (location_id, max_opens_per_member_per_day, max_opens_per_location_per_day, created_at, updated_at)
		 VALUES (?, NULL, NULL, ?, ?)
		 ON CONFLICT (location_id) DO NOTHING`, locationID, t, t); err != nil {
		return Quotas{}, err
	}
	if hasMember {
		if _, err := tx.ExecContext(ctx,
			`UPDATE location_settings SET max_opens_per_member_per_day = ?, updated_at = ? WHERE location_id = ?`,
			nullInt(memberVal), t, locationID); err != nil {
			return Quotas{}, err
		}
	}
	if hasLocation {
		if _, err := tx.ExecContext(ctx,
			`UPDATE location_settings SET max_opens_per_location_per_day = ?, updated_at = ? WHERE location_id = ?`,
			nullInt(locationVal), t, locationID); err != nil {
			return Quotas{}, err
		}
	}
	var q Quotas
	var m, l sql.NullInt64
	if err := tx.QueryRowContext(ctx,
		`SELECT max_opens_per_member_per_day, max_opens_per_location_per_day
		 FROM location_settings WHERE location_id = ?`, locationID).Scan(&m, &l); err != nil {
		return Quotas{}, err
	}
	if m.Valid {
		q.MaxOpensPerMemberPerDay = &m.Int64
	}
	if l.Valid {
		q.MaxOpensPerLocationPerDay = &l.Int64
	}
	return q, tx.Commit()
}

// MemberOpens is one member's successful-open count for the usage breakdown.
type MemberOpens struct {
	UserID     string
	Email      string
	OpensToday int
}

// LocationUsage computes today's successful opens from access_logs over the
// same UTC day window the limiter counts, so UI numbers match enforcement.
func (s *Store) LocationUsage(ctx context.Context, locationID, userID string, dayStart int64) (locationOpens, myOpens int, members []MemberOpens, err error) {
	err = s.db.QueryRowContext(ctx,
		`SELECT
		   count(*) FILTER (WHERE success = 1 AND command = 'open'),
		   count(*) FILTER (WHERE success = 1 AND command = 'open' AND user_id = ?)
		 FROM access_logs WHERE location_id = ? AND ts >= ?`,
		userID, locationID, dayStart).Scan(&locationOpens, &myOpens)
	if err != nil {
		return 0, 0, nil, err
	}
	rows, err := s.db.QueryContext(ctx,
		`SELECT coalesce(al.user_id, ''), coalesce(u.email, ''), count(*)
		 FROM access_logs al LEFT JOIN users u ON u.id = al.user_id
		 WHERE al.location_id = ? AND al.success = 1 AND al.command = 'open' AND al.ts >= ?
		 GROUP BY al.user_id, u.email
		 ORDER BY count(*) DESC
		 LIMIT 50`, locationID, dayStart)
	if err != nil {
		return 0, 0, nil, err
	}
	defer rows.Close()
	for rows.Next() {
		var m MemberOpens
		if err := rows.Scan(&m.UserID, &m.Email, &m.OpensToday); err != nil {
			return 0, 0, nil, err
		}
		members = append(members, m)
	}
	return locationOpens, myOpens, members, rows.Err()
}

func nullStr(p *string) any {
	if p == nil {
		return nil
	}
	return *p
}

func nullFloat(p *float64) any {
	if p == nil {
		return nil
	}
	return *p
}

func nullInt(p *int64) any {
	if p == nil {
		return nil
	}
	return *p
}
